main/src/ca/uwaterloo/flix/runtime/Value.scala

```diff
@@ -418,17 +388,44 @@ object Value {
     case _ => throw new InternalRuntimeException(s"Unexpected non-map value: '$ref'.")
   }
 
+  /////////////////////////////////////////////////////////////////////////////
+  // Equality                                                                //
+  /////////////////////////////////////////////////////////////////////////////
   /**
-    * Casts the given `ref` to a Flix value.
-    */
-  // TODO: This is temporary fix until we figure out a proper way to deal with non-flix values.
-  @inline
-  def cast2flix(ref: AnyRef): AnyRef = ref match {
-    case Tuple1(a: AnyRef) => Value.Tuple(Array(a))
-    case Tuple2(a: AnyRef, b: AnyRef) => Value.Tuple(Array(a, b))
-    case Tuple3(a: AnyRef, b: AnyRef, c: AnyRef) => Value.Tuple(Array(a, b, c))
-    case Tuple4(a: AnyRef, b: AnyRef, c: AnyRef, d: AnyRef) => Value.Tuple(Array(a, b, c, d))
-    case _ => ref
+    * Returns `true` if the values of the two given references `ref1` and `ref2` are equal.
+    */
+  def equal(ref1: AnyRef, ref2: AnyRef): Boolean = ref1 match {
+    case _: Unit.type => ref1 eq ref2
+    case _: java.lang.Boolean => ref1.equals(ref2)
+    case _: java.lang.Character => ref1.equals(ref2)
+    case _: java.lang.Float => ref1.equals(ref2)
+    case _: java.lang.Double => ref1.equals(ref2)
+    case _: java.lang.Byte => ref1.equals(ref2)
+    case _: java.lang.Short => ref1.equals(ref2)
+    case _: java.lang.Integer => ref1.equals(ref2)
+    case _: java.lang.Long => ref1.equals(ref2)
+    case _: java.math.BigInteger => ref1.equals(ref2)
+    case _: java.lang.String => ref1.equals(ref2)
+    case _: Tag =>
+      val v1 = ref1.asInstanceOf[Tag]
+      val v2 = ref2.asInstanceOf[Tag]
```

Comment:
The type system guarantees that you only compare two values of the same type, so if one is tag, the other one must be too.